packages/ui/src/rsc.tsx

```diff
@@ -0,0 +1,61 @@
+import { AsyncLocalStorage } from '@llamaindex/env'
+import { createStreamableUI } from 'ai/rsc'
+import {
+	Settings,
+	CallbackManager,
+	type LlamaIndexEventMaps
+} from '@llamaindex/core/global';
+import type { ComponentType, ReactNode } from 'react';
+
+const uiAsyncLocalStorage = new AsyncLocalStorage()
+
+type UIMap = {
+	[K in keyof LlamaIndexEventMaps]?: ComponentType<{
+		data: LlamaIndexEventMaps[K]
+	}>
+}
+
+export type UIOptions = {
+	initial?: ReactNode
+	on: UIMap
+}
+
+export function run<T>(
+	fn: () => Promise<T>,
```

Comment:
my idea is we have event manager, so we can listen the event and show the loading/result UI to the client side

What do u think

@marcusschiesser